### ISSUE TITLE
Tensors dimensions mismatch in the middle of training

### ISSUE THREAD

**Issue by gireek** [facebookresearch/mmf#510]
I am training on my dataset exactly like Hateful memes with an image and text wanting to classify it. THe only difference is number of classes and I change it in this file: https://github.com/apsdehal/hm_example_mmf/blob/master/configs/models/concat_vl.yaml from 2 to 5. The train.log has a couple of rows till 300/22000 but then the training stops and this error comes:
@apsdehal @vedanuj 

```
Traceback (most recent call last):
  File "/usr/local/bin/mmf_run", line 8, in <module>
    sys.exit(run())
  File "/usr/local/lib/python3.6/dist-packages/mmf_cli/run.py", line 111, in run
    main(configuration, predict=predict)
  File "/usr/local/lib/python3.6/dist-packages/mmf_cli/run.py", line 44, in main
    trainer.train()
  File "/usr/local/lib/python3.6/dist-packages/mmf/trainers/base_trainer.py", line 250, in train
    report = self._forward_pass(batch)
  File "/usr/local/lib/python3.6/dist-packages/mmf/trainers/base_trainer.py", line 274, in _forward_pass
    model_output = self.model(prepared_batch)
  File "/usr/local/lib/python3.6/dist-packages/mmf/models/base_model.py", line 148, in __call__
    model_output = super().__call__(sample_list, *args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 550, in __call__
    result = self.forward(*input, **kwargs)
  File "/content/hm_example_mmf/models/concat_vl.py", line 38, in forward
    [text_features, image_features.squeeze()], dim=1
RuntimeError: Tensors must have same number of dimensions: got 1 and 2

```
Using Google Colab GPU:
python -m torch.utils.collect_env
```
Collecting environment information...
PyTorch version: 1.5.0+cu101
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.5.0+cu101
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.5.0
[pip3] torchvision==0.6.0+cu101
[conda] Could not collect

```
Please guide as in what might be the issue. Thanks

**Comment by vedanuj** [comment]
Can you log the dimension of the tensors before the concat step? There might be an odd sample that has dimension mismatch.

**Comment by gireek** [comment]
```
dir = './data/img/'
a = []
for file_name in os.listdir(dir):
  # print(file_name)
  img = cv2.imread(dir + file_name)
  a.append((img.shape))

print(set(a))
```

I printed the shape of all my images and saved in into a Python set. The value returned was {(256, 256, 3)} All my images were of shape 256,256,3

**Comment by vedanuj** [comment]
I meant can you log the output dimensions of the encoders right before this line in concat_vl.py:

[text_features, image_features.squeeze()], dim=1

**Comment by gireek** [comment]
Do you mean like this:

```
    logger = logging.getLogger(__name__)
    def forward(self, sample_list):
        text = sample_list["text"]
        image = sample_list["image"]
        text_features = torch.nn.functional.relu(
            self.language_module(text)
        )
        image_features = torch.nn.functional.relu(
            self.vision_module(image)
        )
        logger.info(text_features)
        logger.info(image_features)
        logger.info(text_features.shape)
        logger.info(image_features.shape)
```
or is there an MMF logger which goes into train.log?

**Comment by vedanuj** [comment]
Yes. This might be due to the logs not being logged by the master worker. Can you also check the log files for other GPUs? Those will be under the `logs` directory

**Comment by gireek** [comment]
Using Unix commands, I checked there was no other log file created in the system apart from:

```
  1991368     40 -rw-r--r--   1 root     root        37910 Aug 25 12:55 ../content/hm_example_mmf/save/train.log
  1991366     40 -rw-r--r--   1 root     root        37094 Aug 25 12:55 ../content/hm_example_mmf/save/logs/train_2020-08-25T12:38:24.log
```
These 2 log files. So will there be another way to log the dimensions?
@apsdehal @vedanuj

**Comment by apsdehal** [comment]
If you just `print` it should come up in stdout. Your logger would only print to stdout by default. If you want mmf's logger, get it by `logger = logging.getLogger("mmf")`.

**Comment by gireek** [comment]
@apsdehal 

Even the code below did not create a log for the shape before the concat step. This is being changed in this file:https://github.com/apsdehal/hm_example_mmf/blob/master/models/concat_vl.py

```
logger = logging.getLogger("mmf")

@registry.register_model("concat_vl")
class LanguageAndVisionConcat(BaseModel):
    # Not really needed, you can skip this as it doesn't do anything
    # Left here for the brevity purposes
    def __init__(self, config, *args, **kwargs):
        super().__init__(config, *args, **kwargs)
    
    @classmethod
    def config_path(cls):
        # Relative to user dir root
        return "configs/models/concat_vl.yaml"
    
    def build(self):
        self.vision_module = build_image_encoder(self.config.image_encoder)
        self.classifier = build_classifier_layer(self.config.classifier)
        self.language_module = ProjectionEmbedding(**self.config.text_encoder.params)
        self.dropout = torch.nn.Dropout(self.config.dropout)
        self.fusion = torch.nn.Linear(**self.config.fusion.params)

    def forward(self, sample_list):
        text = sample_list["text"]
        image = sample_list["image"]
        text_features = torch.nn.functional.relu(
            self.language_module(text)
        )
        image_features = torch.nn.functional.relu(
            self.vision_module(image)
        )
        logger.info(text_features.shape)
        logger.info(image_features.shape)
        combined = torch.cat(
            [text_features, image_features.squeeze()], dim=1
        )
```

**Comment by apsdehal** [comment]
Can you try direct print, it should print in stdout console? We will have look at logger later.

**Comment by gireek** [comment]
Sorry but print didnt work either. Nothing came in the stdout console.

**Comment by apsdehal** [comment]
Actually, which version of MMF are you on? Can you tell git sha of MMF for you?

**Comment by gireek** [comment]
```
import mmf
print(mmf.__version__)
```
returns 1.0.0rc10

**Comment by apsdehal** [comment]
You need to install from source in your case.
Please update mmf by following these commands:

```
cd ~
git clone https://github.com/facebookresearch/mmf
cd mmf
pip install -e .
```

**Comment by gireek** [comment]
Yes. That worked. It is logging now.
I have kept training batch size as 2
All the logs were like this:
```
2020-08-25T18:30:09 | INFO | hm_example_mmf.models.concat_vl : torch.Size([2, 300])
2020-08-25T18:30:09 | INFO | hm_example_mmf.models.concat_vl : torch.Size([2, 1, 2048])
```
I rechecked that all the text I was sending was non-empty.
but when the error comes the log is:
```
2020-08-25T18:30:09 | INFO | hm_example_mmf.models.concat_vl : torch.Size([1, 300])
2020-08-25T18:30:09 | INFO | hm_example_mmf.models.concat_vl : torch.Size([1, 1, 2048])
```

**Comment by apsdehal** [comment]
Line https://github.com/apsdehal/hm_example_mmf/blob/master/models/concat_vl.py#L38 needs to be changed to

```
            [text_features, image_features.squeeze(dim=1)], dim=1
```

Would you be willing to open a PR?

**Comment by gireek** [comment]
Opened a PR

**Comment by apsdehal** [comment]
Thanks. Can you confirm that this fixes your issue?

**Comment by gireek** [comment]
> Thanks. Can you confirm that this fixes your issue?

Thank you. Training perfectly. Gives great validation accuracy as well.

`!mmf_predict config="configs/experiments/defaults.yaml" model=concat_vl dataset=hateful_memes run_type=test checkpoint.resume_file=./save/models/model_2000.ckpt`
Will this usage of mmf_predict work for prediction on test data?
If I have to write a custom predict function from where to apply all transformations on test image like resizing,etc which MMF does?

**Comment by apsdehal** [comment]
> Will the above usage of mmf_predict work for prediction on test data?

It seems like in your command you are evaluating on test data?

> If I have to write a custom predict function from where to apply all transformations on test image like resizing,etc which MMF does?

Following the classify function in MMBTHMInterface?

**Comment by gireek** [comment]
Hi @apsdehal Thanks for your response. I followed the MMBTHMInterface for loading my own checkpoint and I loaded it into model like you do in your demo. The only problem is in the concat_vl version there is no image_processor in the yaml file. But this line here: https://github.com/facebookresearch/mmf/blob/master/mmf/models/interfaces/mmbt.py#L70 asks for the image to go through a processor. How should I counter this?
